runtime-sdk/modules/contracts/src/abi/oasis/memory.rs

```diff
@@ -152,4 +172,67 @@ impl<Cfg: Config> OasisV1<Cfg> {
         let data = cbor::to_vec(data);
         Self::allocate_and_copy(instance, &data)
     }
+
+    /// Serializes the given type into CBOR, allocates a chunk of memory inside the WASM instance
+    /// and copies the region and serialized data into it. Returns a pointer to the serialized region.
+    ///
+    /// This method is useful when you need to a pointer to the region of the serialized data,
+    /// since it avoids an additional allocation for the region itself as it pre-allocates it with the data.
+    /// This is an optimized version of calling `serialize_and_allocate` followed by `allocate_region`
+    /// which does two separate allocations.
+    pub fn serialize_and_allocate_as_ptr<C, T>(
+        instance: &wasm3::Instance<'_, '_, ExecutionContext<'_, C>>,
+        data: T,
+    ) -> Result<u32, RegionError>
+    where
+        C: Context,
+        T: cbor::Encode,
+    {
+        let data = cbor::to_vec(data);
+        // Allocate enough for the data and the serialized region.
+        let outer = Self::allocate(instance, data.len() + 8)?;
+        // First 8 bytes are reserved for the region itself. Inner is the region
+        // for the actual data.
+        let inner = Region {
+            offset: outer.offset + 8,
+            length: outer.length - 8,
+        };
+
+        instance
+            .runtime()
+            .try_with_memory(|mut memory| -> Result<(), RegionError> {
+                inner.copy_from_slice(&mut memory, &data)?;
+
+                let dst = &mut memory.as_slice_mut()[outer.offset..outer.offset + 8];
```

Comment:
This may panic due to a TODO in `allocate` about early validation if the guest returns an invalid offset. Let's add that validation to `allocate` (currently the check is only performed in region's `copy_from_slice`).

Reply:
added